### PR TITLE
dev to main for 2024 release

### PR DESCRIPTION
**Pull request by brtietz**
Major changes include:

- Split offshore wind into fixed and floating
- Add optional variables for which tax credit a technology is taking
- Use these variables to allow PV+Battery to take different credits for the PV and battery components
- Additional tech detail changes for Nuclear and Natural Gas to match the new ATB
- Updated financial assumptions

### CODE REVIEW

File: lcoe_calculator/base_processor.py [hunk 1]
```diff
@@ -524,4 +555,37 @@ def _calc_lcoe(self):
...
+        return df_lcoe
+
+    def _get_tax_credit_case(self):
+        """
+        Uses ptc and itc data from the tech sheet to determine which tax credits are active for the
+        current financial case and tax credit case
+
+        @returns String, one of "None", "PTC", "ITC", "ITC + PTC"
+        """
+        if self.has_tax_credit:
```
Comment (mikebannis) [inline lcoe_calculator/base_processor.py:567]: Suggest converting this to a guard clause do reduce indenting: 
```
if not self.has_tax_credit:
    return "None"
```

File: lcoe_calculator/base_processor.py [hunk 2]
```diff
@@ -524,4 +555,37 @@ def _calc_lcoe(self):
...
+            itc_sum = np.sum(itc)
+
+            if ptc_sum > 0 and itc_sum > 0:
+                return "PTC + ITC"
+            elif ptc_sum > 0:
+                return "PTC"
+            elif itc_sum > 0:
+                return "ITC"
+            else:
+                return "None"
```
Comment (mikebannis) [inline lcoe_calculator/base_processor.py:588]: Suggest dropping elif/else and converting to pure if:
```
if ... : 
    return ...

if ...  : 
   return ...

return ...
```

File: lcoe_calculator/tech_processors.py [hunk 3]
```diff
@@ -88,26 +103,107 @@ class UtilityPvPlusBatteryProc(TechProcessor):
...
+            self.df_fin = extractor.get_fin_assump()
+
+        if self.has_wacc:
+            print('\tLoading WACC data')
+            self.df_wacc, self.df_just_wacc = extractor.get_wacc(self.wacc_name)
+
+        print('\tDone loading data')
+        return extractor
+
+    def _get_tax_credit_case(self):
```
Comment (mikebannis) [inline lcoe_calculator/tech_processors.py:179]: Please add doc string to explain diff from BaseProcessor

File: lcoe_calculator/tech_processors.py [hunk 4]
```diff
@@ -88,26 +103,107 @@ class UtilityPvPlusBatteryProc(TechProcessor):
...
+        ('PV-only Capacity Factor (%)','df_pvcf')
     ]
 
+    def __init__(
+        self,
+        data_workbook_fname: str,
+        case: str = MARKET_FIN_CASE,
+        crp: CrpChoiceType = 30,
+        tcc: str = "PV PTC and Battery ITC",
+        extractor: type[AbstractExtractor] = PVBatteryExtractor
```
Comment (mikebannis) [inline lcoe_calculator/tech_processors.py:115]: type should be capitalized, but see other comments

**Comment by mikebannis** [comment]
A couple questions/comments about UtilityPVPlusBatteryProc. 

1. It uses a custom extractor, but that extractor is not being used in `extract_test_data.py`. Is UtilityPVPlusBattery being adequately tested?
2. Since it uses a custom extractor, I'm tempted to say the extractor argument should be removed from the `__init__()` definition on line 115 of `tech_processors.py`, and just have it be hardwired as a class attribute. I don't think that will break anything in the repo. Alternatively, allow the `extractor` argument if necessary, and ignore it. That feels dirty and should be documented.

File: lcoe_calculator/tech_processors.py [hunk 5]
```diff
@@ -65,15 +80,15 @@ class ResPvProc(TechProcessor):
...
     tech_name = 'Utility-Scale PV-Plus-Battery'
     tech_life = 30
     sheet_name = 'Utility-Scale PV-Plus-Battery'
     num_tds = 10
     default_tech_detail = 'PV+Storage - Class 5'
-    dscr = 1.3
+    dscr = 1.25
 
     GRID_ROUNDTRIP_EFF = 0.85 # Roundtrip Efficiency (Grid charging)
     CO_LOCATION_SAVINGS = 0.9228
```
Comment (mikebannis) [inline lcoe_calculator/tech_processors.py:94]: Please document what this value represents. CO === Colorado?

**Comment by mikebannis** [comment]
Looks good to me other than comments above

**Comment by brtietz** [comment]
> A couple questions/comments about UtilityPVPlusBatteryProc.
> 
>     1. It uses a custom extractor, but that extractor is not being used in `extract_test_data.py`. Is UtilityPVPlusBattery being adequately tested?

Good question. The reason for the custom extractor is to allow for specifying the tax credit case during processing. So we're adequately testing the PTC+ITC case, but not the "ITC only" case. We'd have to do some upgrades to extract_test_data.py and test_lcoe_calculator.py to cover both cases.

>     2. Since it uses a custom extractor, I'm tempted to say the extractor argument should be removed from the `__init__()` definition on line 115 of `tech_processors.py`, and just have it be hardwired as a class attribute. I don't think that will break anything in the repo. Alternatively, allow the `extractor` argument if necessary, and ignore it. That feels dirty and should be documented.

Would this also require a change to base_processor.py line 107?

Thoughts on handling these in this PR vs making a separate issue? I can see some advantages to merging a little sooner in case users try to use main with the new spreadsheet, but happy to wait to tag the release until this is resolved.

File: lcoe_calculator/tech_processors.py (hunk 4 above)
Comment (mikebannis) [reply lcoe_calculator/tech_processors.py:115]: New idea: change this to `extractor: Type[PVBatteryExtractor] = PVBatteryExtractor`. It's painfully obvious it needs to be the special extractor or a derivative, and no changes needed anywhere else.

**Pull request merged**